### PR TITLE
Revert "Revert "Negative indices for list nodes, and list insertion and deletion""

### PR DESCRIPTION
**Pull request by Steell**
Reverts DynamoDS/Dynamo#3307

@lukechurch for review.

### CODE REVIEW

**Comment by Steell** [comment]
ping @aparajit-pratap @ikeough @pboyer or @lukechurch

**Comment by aparajit-pratap** [comment]
LGTM!

**Comment by ikeough** [comment]
@lukechurch You were originally set as the review on this. I want to make sure this doesn't run counter to any assumptions in the language. PTAL.

**Comment by pboyer** [comment]
Ping @lukechurch

File: src/Libraries/CoreNodes/List.cs [hunk 1]
```diff
@@ -260,6 +260,38 @@ public static IList AddItemToEnd(object item, IList list)
...
+        ///     Inserts an item into a list at a specified index.
+        /// </summary>
+        /// <param name="item">Item to be inserted.</param>
+        /// <param name="list">List to insert in to.</param>
+        /// <param name="index">Index to insert item at.</param>
+        /// <search>insert,add,item</search>
+        public static IList InsertItemAtIndex(object item, IList list, int index)
+        {
+            while (index < 0)
+                index += list.Count;
```
Comment (lukechurch) [inline src/Libraries/CoreNodes/List.cs]: I don't believe this is the correct behaviour.

{0, 2}[3] fails with an out of bounds error
  so
{0, 2}[-3] should also fail.

**Comment by pboyer** [comment]
Hi all, We've added this API so I'm closing this.

**Pull request closed without merging**